### ISSUE TITLE
UnsupportedApiVersion from route without ApiVersioning

### ISSUE THREAD

**Issue by thiagomajesk** [Microsoft/aspnet-api-versioning#174]
Hi, I'm having trouble with routes that doesn't use the ApiVersioning attribute. Considering the controllers bellow:

```c#
[ApiVersion("1.0"), RequireHttps]
[Route("api/v{version:apiVersion}/[controller]")]
public class FooController : Controller
{
    /* [...] */
}
```
I've configured it like this: 
`services.AddApiVersioning(o => o.AssumeDefaultVersionWhenUnspecified = true);`. 
And then, accessing a `bar/create` route in the `BarController` that does not use the ApiVersioning attribute, I get `UnsupportedApiVersion` error. I have assumed that the ApiVersioning attribute would only override the route for the controller I specified, but that doesn't seem to be true.

**Comment by commonsensesoftware** [comment]
API versioning doesn't really get involved with routing. A design goal has always been to not force developers to learn anything new about routing in ASP.NET. The `[ApiVersion]` attribute is purely metadata. You can also use conventions.

Once you opt into API versioning, there is no such thing as an _unversioned API_ anymore. Controllers that don't have any attributes still (and conceptually always did) have an API version. The assumed value in these cases is `ApiVersioningOptions.DefaultAPiVersion`, which has the initial value of `1.0`. The primary reason for this behavior is so that existing APIs (e.g. controllers) can be _grandfathered_ in without change.

There an option to use `[ApiVersionNeutral]`, but you should know that this doesn't mean _unversioned_. This means exactly the opposite - _I accept any API version_. This is the closest behavior you can get to _unversioned_.

The `ApiVersioningOptions.AssumeDefaultVersionWhenUnspecified` allows for a client to make a request without specifying an API version. When this scenario happens, the API version used will be based on the behavior of the configured `ApiVersioningOptions.ApiVersionSelector`. The default configuration always selects `ApiVersioningOptions.DefaultAPiVersion`. All of the other out-of-the-box implementations also fallback to this value when other selections do not yield a result.

You may be thinking you have configured all of this, so why isn't `bar/create` working. This has everything to do with routing. Since you are routing by using a URL segment, that segment **must** be specified in order for ASP.NET to route to it. There is no other way, that I know of, that enables you to have a default value in the middle of a URL (though you can at the trail end). I've outlined this issue in detail in the [Known Limitations - URL Path Segment Routing with a Default API Version](../wiki/Known-Limitations#url-path-segment-routing-with-a-default-api-version) topic.

In short, the only way that you can make `bar/create` work is to have two route templates (including the rest of your existing configuration):

```c#
[Route( "[controller]") ]
[Route( "v{version:apiVersion}/[controller]" )]
public class BarController : Controller
{
  [HttpPost( "create" )]
  public IActionResult Create( Bar bar ) => Status( 201 );
}
```

If the **BarController** doesn't care which API version is requested, then you can use:

```c#
[ApiVersionNeutral]
[Route( "[controller]") ]
public class BarController : Controller
{
  [HttpPost( "create" )]
  public IActionResult Create( Bar bar ) => Status( 201 );
}
```

I hope that helps. Let me know if you have more questions.

**Comment by thiagomajesk** [comment]
@commonsensesoftware 
> Once you opt into API versioning, there is no such thing as an unversioned API anymore.  

Its' weird that this is the default behaviour though. I still think this should be opt-in per controller.

So, you would say that the prefered way to do that without messing up with routes (both templates and query strings) would be using the versioning through [media type](https://github.com/Microsoft/aspnet-api-versioning/wiki/Versioning-by-Media-Type)?

**Comment by thiagomajesk** [comment]
@commonsensesoftware No problem at all. Thanks for the feedback and thanks for the explanation btw :)

About the issue regrading version-neutral routes: I'll try to make a repro of this problem and post it here, it may take a few days though, since I'm about to release my web app.

**Comment by commonsensesoftware** [comment]
Sounds like you're on your way. Should you have a repro or something else you'd like to share, feel free to do so. Sample contributions from the community help a lot with ironing out edge cases. Should something pop up or you find the issue isn't resolved, we can reopen this issue. Thanks.
